File: scripts/prepare-lora-training.ts
/**
 * Prepare LoRA Training Images
 * 
 * This script collects all Marshall training images and creates a zip file
 * ready for LoRA training.
 * 
 * Run with: npx tsx scripts/prepare-lora-training.ts
 */

import * as fs from "fs";
import * as path from "path";
import archiver from "archiver";

const PROJECT_ROOT = path.join(__dirname, "..");
const OUTPUT_DIR = path.join(PROJECT_ROOT, "scripts", "temp");
const ZIP_PATH = path.join(OUTPUT_DIR, "marshall-training-images.zip");

// All directories containing training images
const IMAGE_DIRS = [
  path.join(PROJECT_ROOT, "public", "consistency_test"),
  path.join(PROJECT_ROOT, "public", "assets"),
  path.join(PROJECT_ROOT, "public", "casting"),
];

async function prepareTrainingImages() {
  console.log("📦 Preparing LoRA training images...\n");

  // Create temp directory
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }

  // Collect all image files
  const imageFiles: string[] = [];
  const imageExtensions = [".png", ".jpg", ".jpeg", ".webp"];

  for (const dir of IMAGE_DIRS) {
    if (!fs.existsSync(dir)) {
      console.warn(`⚠️  Directory not found: ${dir}`);
      continue;
    }

    const files = fs.readdirSync(dir);
    for (const file of files) {
      const ext = path.extname(file).toLowerCase();
      if (imageExtensions.includes(ext)) {
        const fullPath = path.join(dir, file);
        imageFiles.push(fullPath);
        console.log(`✅ Found: ${file}`);
      }
    }
  }

  if (imageFiles.length === 0) {
    console.error("❌ No image files found!");
    process.exit(1);
  }

  console.log(`\n📊 Total images: ${imageFiles.length}`);
  console.log("📦 Creating zip file...\n");

  // Create zip file
  return new Promise<string>((resolve, reject) => {
    const output = fs.createWriteStream(ZIP_PATH);
    const archive = archiver("zip", { zlib: { level: 9 } });

    output.on("close", () => {
      const sizeMB = (archive.pointer() / 1024 / 1024).toFixed(2);
      console.log(`✅ Zip file created: ${ZIP_PATH}`);
      console.log(`📦 Size: ${sizeMB} MB`);
      console.log(`📁 Contains ${imageFiles.length} images\n`);
      resolve(ZIP_PATH);
    });

    archive.on("error", (err) => {
      reject(err);
    });

    archive.pipe(output);

    // Add all images to zip
    for (const imagePath of imageFiles) {
      const fileName = path.basename(imagePath);
      archive.file(imagePath, { name: fileName });
    }

    archive.finalize();
  });
}

// Run the script
prepareTrainingImages()
  .then((zipPath) => {
    console.log("✨ Preparation complete!");
    console.log(`\n📦 Zip file ready: ${zipPath}`);
    console.log("\nNext step: Run training with:");
    console.log("   npx tsx scripts/train-lora-fal.ts\n");
  })
  .catch((error) => {
    console.error("❌ Error:", error);
    process.exit(1);
  });

File: scripts/prepare-lora-training.py
"""
Prepare LoRA Training Images

This script collects all Marshall training images and creates a zip file
ready for LoRA training.

Run with: python scripts/prepare-lora-training.py
"""

import os
import zipfile
from pathlib import Path

# Get paths
SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent
OUTPUT_DIR = PROJECT_ROOT / "scripts" / "temp"
ZIP_PATH = OUTPUT_DIR / "marshall-training-images.zip"

# All directories containing training images
IMAGE_DIRS = [
    PROJECT_ROOT / "public" / "consistency_test",
    PROJECT_ROOT / "public" / "assets",
    PROJECT_ROOT / "public" / "casting",
]

def main():
    print("📦 Preparing LoRA training images...\n")

    # Create temp directory
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    # Collect all image files
    image_files = []
    image_extensions = {".png", ".jpg", ".jpeg", ".webp"}

    for dir_path in IMAGE_DIRS:
        if not dir_path.exists():
            print(f"⚠️  Directory not found: {dir_path}")
            continue

        for file_path in dir_path.iterdir():
            if file_path.is_file() and file_path.suffix.lower() in image_extensions:
                image_files.append(file_path)
                print(f"✅ Found: {file_path.name}")

    if not image_files:
        print("❌ No image files found!")
        exit(1)

    print(f"\n📊 Total images: {len(image_files)}")
    print("📦 Creating zip file...\n")

    # Create zip file
    with zipfile.ZipFile(ZIP_PATH, "w", zipfile.ZIP_DEFLATED) as zipf:
        for image_path in image_files:
            # Add file to zip with just the filename (not full path)
            zipf.write(image_path, image_path.name)
            print(f"   Added: {image_path.name}")

    # Get file size
    size_mb = ZIP_PATH.stat().st_size / 1024 / 1024

    print(f"\n✅ Zip file created: {ZIP_PATH}")
    print(f"📦 Size: {size_mb:.2f} MB")
    print(f"📁 Contains {len(image_files)} images\n")
    print("✨ Preparation complete!")
    print("\nNext step: Run training with:")
    print("   python scripts/train-lora.py\n")

if __name__ == "__main__":
    main()
